README.md
# BitBox02 JavaScript library

The API is currenly unstable. Expect frequent breaking changes until we start tagging versions.

The JavaScript library is compiled from
[bitbox02-api-go](https://github.com/digitalbitbox/bitbox02-api-go) using GopherJS.

## Integration

To talk to the device from a browser using this API, you will need the BitBoxBridge: https://github.com/digitalbitbox/bitbox-bridge
To integrate the BitBox02 into your application, your domain will need to be whitelisted in the Bridge. To do so, please submit a Pull Request or an Issue in the [bitbox-bridge](https://github.com/digitalbitbox/bitbox-bridge) repo.

### Install:
```sh
$ npm install bitbox02-api
```

### Import
```javascript
import { BitBox02API, getDevicePath} from 'bitbox02-api';
```

### Initialize
To get the device path, the BitBoxBridge needs to be running

```javascript
const devicePath = await getDevicePath();
const BitBox02 = new BitBox02API(devicePath);
```

### Connect
The `BitBox02API.connect()` method takes 5 arguments:
```javascript
/**
 * @param showPairingCb Callback that is used to show pairing code. Must not block.
 * @param userVerify Promise that should resolve once the user wants to continue.
 * @param handleAttastionCb Callback that should print "attestation failed". Must not block.
 * @param onCloseCb Callback thats called when the websocket connection is closed.
 * @param setStatusCb Callback that lets the API set the status received from the device.
 * @return Promise that will resolve once the pairing is complete.
 */
connect (showPairingCb, userVerify, handleAttestationCb, onCloseCb, setStatusCb)
```

For more detailed example see the [Sample integration](https://github.com/digitalbitbox/bitbox02-api-js/blob/master/README.md#sample-integration) below.

### Methods and functions

#### ethGetRootPubKey: Get Ethereum Root Pub Key
Get eth xpub for a given coin and derivation path

```javascript
/**
 * @param keyPath account keypath in string format
 * Currently only two keypaths are supported:
 * - `m/44'/60'/0'/0` for mainnet and
 * - `m/44'/1'/0'/0`  for Rinkeby and Ropsten testnets
 */
const rootPub = await BitBox02.ethGetRootPubKey(keyPath: string);
```

#### ethSignTransaction: Sign Ethereum transaction
To sign Etehreum transactions, we recommend using the `Transaction` type provided by the `ethereumjs` library https://github.com/ethereumjs/ethereumjs-tx/blob/master/src/transaction.ts and passing this transaction data through our `sanitizeEthTransactionData(sigData)` function to prepare the tx for the API:

```javascript
/** @param sigData should include the following where `tx` is the `Transaction` from `ethereumjs`:
  *
  * const sigData = {
  *     path: keypath     // full Ethereum account keypath string e.g. m/44'/60'/0'/0/0
  *     recipient: tx.to, // Buffer(Uint8Array(20))
  *     tx: {
  *       value           // hex
  *       data            // hex
  *       chainId         // number
  *       nonce           // hex
  *       gasLimit        // hex
  *       gasPrice        // hex
  *      },
  *     data: tx.data // Buffer(Uint8Array)
  *   }
  * ```
  */
function sanitizeEthTransactionData(sigData)
```

Then to send the data to the device and get the signature bytes:
```javascript
import { sanitizeEthTransactionData } from 'bitbox02-api';

const sanitizedData = sanitizeEthTransactionData(sigData);
const result = await BitBox02.ethSignTransaction(sanitizedData);
```

#### ethSignMessage: Sign Ethereum messages
```javascript
/** @param msgData is an object including the account id and the message as bytes/Buffer:
  *
  * const msgData = {
  *     account    // number, account number in the ETH keypath e.g. m/44'/60'/0'/0/<id>
  *     message    // Buffer
  *   }
  */
const result = await BitBox02.ethSignMessage(msgData);
```

#### ethDisplayAddress: Display Ethereum address on BitBox02 screen for verification
```javascript
/** 
 * @param id is a number; account number in the ETH keypath e.g. m/44'/60'/0'/0/<id>:
 */
await BitBox02.ethDisplayAddress(id)
```

#### Check if product is supported
To use with Ethereum, the user needs a BB02 Multi and not the Bitcoin Only edition. You can use for the following check:
```javascript
import { api } from 'bitbox02-api';

BitBox02.fw.Product() === api.common.Product.BitBox02Multi
```

## Sample integration
This is a sample BitBox02Wallet class integration for connecting to the BitBox02 device using the BitBoxBridge and this JS API

```javascript
import {
  BitBox02API,
  getDevicePath,
  api,
  sanitizeEthTransactionData
} from 'bitbox02-api';

class BitBox02Wallet {
  constructor(logout) {  // You can provide the `logout` callback of your application in the constructor
    this.logout = logout;
    this.status = undefined;
    this.pairingConfirmed = false;
  }

  async connect() {
    const devicePath = await getDevicePath();
    this.BitBox02 = new BitBox02API(devicePath);
  }

  async init(keyPath) {
    await this.BitBox02.connect(

      /** @param showPairingCb
       *  Store the pairing code on the class instance. Show this to the user to compare with code
       *  on the device when `this.status === 'unpaired'`
       */
      pairingCode => {
        this.pairingCode = pairingCode;
      },

      /** @param userVerify
       *  Store the Promise's `resolve` on the class instance to call when the user clicks the corresponding button
       *  in your application after confirming the pairing on device
       */
      async () => {
        return new Promise(resolve => {
          this.pairingConfirmed = true;
          this.pairingConfirmationResolve = resolve;
        });
      },

      /** @param handleAttastionCb
      *  Store the attestation result on the class instance. If attestation fails, the user might have a fake device.
      *  Handle this condition below.
      */
      attestationResult => {
        this.attestation = attestationResult;
      },

      /** @param onCloseCb
      *  Log the user out of your application when device is unplugged/the websocket closes.
      *  Here we use the `logout` function provided in the constructor as the callback.
      */
      () => {
        this.logout();
      },

      /** @param setStatusCb
      *  Store the status on the class instance to take appropriate actions based on status.
      *  All possible status can be found here: https://github.com/digitalbitbox/bitbox02-api-go/blob/master/api/firmware/status.go
      */
      status => {
        this.status = status;
      }
    );

    if (this.BitBox02.fw.Product() !== api.common.Product.BitBox02Multi) {
      throw new Error('Unsupported device');
    }

    // Handle attestattion failure
    if (!this.attestation) {
      errorHandler('Attestation failed');
    }

    const rootPub = await this.BitBox02.ethGetRootPubKey(keyPath);
    // Derive accounts from xpub ..., e.g. `this.hdKey = HDKey.fromExtendedKey(rootPub);`

  }

```

## Develop locally

To compile:

```sh
$ make dockerinit
$ make dockercompile
```

To run a demo: `$ make servedemo` and visit `localhost:8000`


CHANGELOG.md
# 0.2.0

- Add CHANGELOG
- Update README
- Add AsyncBTCAddressSimple, AsyncBTCSignSimple
- Add `getCoinFromChainId()`, `getPathFromString()`, and `getCoinFromPath()` eth utils to handle mainnet and testnet keypaths and coins correctly. Currently only two keypaths are supported:
    - `m/44'/60'/0'/0` for mainnet and
    - `m/44'/1'/0'/0`  for Rinkeby and Ropsten testnets
  

# 0.1.0

- Publish first API version


src/eth-utils.js
import { firmwareAPI, HARDENED } from './bitbox02';

const getCoinFromChainId = chainId => {
    switch(chainId) {
        case 1:
            return firmwareAPI.messages.ETHCoin.ETH;
        case 3:
            return firmwareAPI.messages.ETHCoin.RopstenETH;
        case 4:
        case 42:
            return firmwareAPI.messages.ETHCoin.RinkebyETH;
        default:
            throw new Error('Unsupported network');
    }
}

/**
 * @param keypathString keypath in string format e.g. m/44'/1'/0'/0
 * @returns keypath as array e.g. [2147483692, 2147483649, 2147483648, 0]
 */
export const getKeypathFromString = keypathString => {
    let hardened = false;
    const levels = keypathString.toLowerCase().split('/');
    if (levels[0] !== 'm') throw new Error('Invalid keypath');
    return levels.flatMap(level => {
        if (level === 'm' || level === '') {
            return [];
        }
        if (level.substring(level.length - 1) === "'") {
            hardened = true
        }
        level = parseInt(level);
        if (isNaN(level) || 
            level < 0 ||
            (!hardened && level >= HARDENED) ||
            (hardened && level >= HARDENED * 2)
           ) {
            throw new Error('Invalid keypath');
        }
        if (hardened) level += HARDENED;
        return level;
    })
}

/**
 * @param keypathArray keypath as an array of ints e.g. [2147483692, 2147483649, 2147483648, 0]
 * FIXME: This is a slight hack until the device is provided with the network by the integrating service
 * The only noticeable consequence is that when using the Rinkeby testnet, the user would see 'Ropsten' on device
 * @returns ETHCoin.ETH for mainnet ([44, 60]) and ETHCoin.RopstenETH for testnets ([44, 1])
 */
export const getCoinFromKeypath = keypathArray => {
    if (keypathArray[0] !== 44 + HARDENED) {
        throw new Error('Invalid keypath');
    }
    switch(keypathArray[1]) {
        case 60 + HARDENED:
            return firmwareAPI.messages.ETHCoin.ETH;
        case 1 + HARDENED:
            return firmwareAPI.messages.ETHCoin.RopstenETH;
        default:
            throw new Error('Invalid keypath');
    }
}

/**
 * Sanitizes signature data provided by the 'ethereumjs' library's Transaction type
 * https://github.com/ethereumjs/ethereumjs-tx/blob/master/src/transaction.ts
 * and returns them in the format needed by BB02's AsyncETHSign
 *
 * @param sigData should include the following:
 *
 * ```
 * const signatureData = {
 *     account: id,      // id: number, account number in the ETH keypath m/44'/60'/0'/0/<id>
 *     recipient: tx.to, // Buffer(Uint8Array(20))
 *     tx: {
 *       value           // hex
 *       data            // hex
 *       chainId         // number
 *       nonce           // hex
 *       gasLimit        // hex
 *       gasPrice        // hex
 *      },
 *     data: tx.data // Buffer(Uint8Array)
 *   }
 * ```
 */
export const sanitizeEthTransactionData = sigData => {
    try {
        let sanitizedData = {};
        sanitizedData.nonce = 0;
        sanitizedData.value = '0';
        sanitizedData.coin = getCoinFromChainId(sigData.tx.chainId);
        sanitizedData.keypath = getKeypathFromString(sigData.keypath);
        if (sigData.tx.nonce) {
            sanitizedData.nonce = parseInt(sigData.tx.nonce, 16)
        }
        sanitizedData.gasPrice = parseInt(sigData.tx.gasPrice, 16).toString();
        sanitizedData.gasLimit = parseInt(sigData.tx.gasLimit, 16);
        sanitizedData.recipient = new Buffer(sigData.recipient);
        if (sigData.tx.value) {
            sanitizedData.value = parseInt(sigData.tx.value, 16).toString();
        }
        sanitizedData.data = new Buffer(sigData.data);
        sanitizedData.chainId = sigData.tx.chainId;
        return sanitizedData;
    } catch (e) {
        throw new Error('ethTx data sanitization failed: ', e);
    }
}


src/bitbox02.js
import './bitbox02-api-go.js';

import { getPathFromString, getCoinFromPath } from './eth-utils';

export const api = bitbox02;
export const firmwareAPI = api.firmware;
export const HARDENED = 0x80000000;

function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

// Will try for 1 second to find a device through the bridge.
export async function getDevicePath() {
    const attempts = 10;
    for (let i=0; i<attempts; i++){
        let response;
        try {
            response = await fetch("http://localhost:8178/api/v1/devices", {
                method: 'GET',
                headers: {},
            });
        } catch {
            throw new Error('BitBoxBridge not found');
        }
        const devices = (await response.json()).devices;
        if (devices.length !== 1) {
            await sleep(100);
            continue;
        }
        const devicePath = devices[0].path;
        return devicePath;
    }
    throw new Error("Expected one BitBox02");
}

function promisify(f) {
    return function(...args) {
        return new Promise((resolve, reject) => f(
            (...results) => {
                const err = results.pop();
                if (err !== null) {
                    return reject(err);
                }
                return resolve(...results);
            },
            ...args));
    };
}

export class BitBox02API {
    constructor(devicePath)  {
        this.devicePath = devicePath;
        this.opened = false;
    }

    // @param showPairingCb Callback that is used to show pairing code. Must not block.
    // @param userVerify Promise that should resolve once the user wants to continue.
    // @param handleAttastionCb Callback that should print "attestation failed". Must not block.
    // @param onCloseCb Callback thats called when the websocket connection is closed.
    // @return Promise that will resolve once the pairing is complete.
    connect (showPairingCb, userVerify, handleAttestationCb, onCloseCb, setStatusCb) {
        const self = this;
        self.opened = true;
        return new Promise((resolve, reject) => {
            function onWrite(bytes) {
                if (self.socket.readyState != WebSocket.OPEN) {
                    console.log("Error, trying to write to closed socket");
                    return;
                }
                self.socket.send(bytes);
            }

            self.socket = new WebSocket("ws://127.0.0.1:8178/api/v1/socket/" + self.devicePath)
            self.socket.binaryType = 'arraybuffer'
            self.socket.onopen = async function (event) {
                try {
                    self.fw = firmwareAPI.New(onWrite);

                    // Turn all Async* methods into promises.
                    for (const key in self.fw.js) {
                        if (key.startsWith("Async")) {
                            self.fw.js[key] = promisify(self.fw.js[key]);
                        }
                    }

                    self.fw.SetOnEvent(ev => {
                        if (ev === firmwareAPI.Event.StatusChanged && self.fw) {
                            setStatusCb(self.fw.Status());
                        }
                        if (ev === firmwareAPI.Event.StatusChanged && self.fw.Status() === firmwareAPI.Status.Unpaired) {
                            const [channelHash] = self.fw.ChannelHash();
                            showPairingCb(channelHash);
                        }
                        if (ev === firmwareAPI.Event.AttestationCheckDone) {
                            handleAttestationCb(self.fw.Attestation());
                        }
                        if (ev === firmwareAPI.Event.StatusChanged && self.fw.Status() === firmwareAPI.Status.RequireFirmwareUpgrade) {
                            self.socket.close();
                            reject('Firmware upgrade required');
                        }
                        if (ev === firmwareAPI.Event.StatusChanged && self.fw.Status() === firmwareAPI.Status.RequireAppUpgrade) {
                            self.socket.close();
                            reject('Unsupported firmware');
                        }
                        if (ev === firmwareAPI.Event.StatusChanged && self.fw.Status() === firmwareAPI.Status.Uninitialized) {
                            self.socket.close();
                            reject('Uninitialized');
                        }
                    });

                    await self.fw.js.AsyncInit();
                    switch(self.fw.Status()) {
                        case firmwareAPI.Status.PairingFailed:
                            self.socket.close();
                            throw new Error("Pairing rejected");
                        case firmwareAPI.Status.Unpaired:
                            await userVerify()
                            await self.fw.js.AsyncChannelHashVerify(true);
                            break;
                        case firmwareAPI.Status.Initialized:
                            // Pairing skipped.
                            break;
                        default:
                            throw new Error("Unexpected status: " + self.fw.Status() + "," + firmwareAPI.Status.Unpaired);
                    }

                    resolve();
                } catch(err) {
                    reject(err);
                }
            }
            self.socket.onerror = function(event) {
                reject("Your BitBox02 is busy");
            }
            self.socket.onmessage = function(event) {
                self.fw.js.OnRead(new Uint8Array(event.data));
            }
            self.socket.onclose = function(event) {
                onCloseCb();
            }
        });
    }

    // @return the eth xpub for a given coin and derivation path
    async ethGetRootPubKey(path) {
        const pathArray = getPathFromString(path);
        const coin = getCoinFromPath(pathArray);
        const xpub = await this.fw.js.AsyncETHPub(
            coin,
            pathArray,
            firmwareAPI.messages.ETHPubRequest_OutputType.XPUB,
            false,
            new Uint8Array()
          );
        return xpub
    };

    // Displays the address of the provided ethereum account on device screen
    async ethDisplayAddress(path) {
        const pathArray = getPathFromString(path);
        // FIXME: see def of `getCoinFromPath()`, since we use the same keypath for Ropsten and Rinkeby,
        // the title for Rinkeby addresses will show 'Ropsten' instead
        const coin = getCoinFromPath(pathArray);
        this.fw.js.AsyncETHPub(
            coin,
            pathArray,
            firmwareAPI.messages.ETHPubRequest_OutputType.ADDRESS,
            true,
            new Uint8Array()
          );
    };

    // Signs an ethereum transaction on device
    // @return the signature from the device as bytes
    async ethSignTransaction(sigData) {
        try {
            const sig = await this.fw.js.AsyncETHSign(
                sigData.coin,
                sigData.path,
                sigData.nonce,
                sigData.gasPrice,
                sigData.gasLimit,
                sigData.recipient,
                sigData.value,
                sigData.data
            );
            const chainId = sigData.chainId;
            const vOffset = chainId * 2 + 8;
            const result = {
                r: new Buffer(sig.slice(0, 0 + 32)),
                s: new Buffer(sig.slice(0 + 32, 0 + 32 + 32)),
                v: new Buffer([parseInt(sig.slice(64), 16) + 27 + vOffset])
            };
            return result;
        } catch (err) {
                if (firmwareAPI.IsErrorAbort(err)) {
                throw new Error('User abort');
            } else {
                throw new Error(err.Message);
            }
        }
    };

    async ethSignMessage(msgData) {
        try {
          const sig = await this.fw.js.AsyncETHSignMessage(
            firmwareAPI.messages.ETHCoin.ETH,
            [44 + HARDENED, 60 + HARDENED, 0 + HARDENED, 0, msgData.account],
            msgData.message
          );

          const result = {
              r: new Buffer(sig.slice(0, 0 + 32)),
              s: new Buffer(sig.slice(0 + 32, 0 + 32 + 32)),
              v: new Buffer([parseInt(sig.slice(64), 16) + 27])
          };
          return result;
        } catch(err) {
          if (firmwareAPI.IsErrorAbort(err)) {
              throw new Error('User abort');
          } else {
              throw new Error(err.Message);
          }
        }
    }

    // @return True if the connection has been opened and successfully established.
    connectionValid() {
        return this.opened && this.socket.readyState == WebSocket.OPEN;
    }

    // @return False if connection wasn't opened
    close() {
        if (!this.connectionValid()) {
            return false;
        }
        this.socket.close();
        return true;
    }

    // Use the return value of this function to communicate with device
    // @return Undefined if connection wasn't opened
    firmware() {
        if (!this.connectionValid()) {
            return undefined;
        }
        return this.fw;
    }
}
